### PR TITLE
[pulsetracker-0001] Add plugin Class

### PR DESCRIPTION
**Pull request by jcantosg**
* Se ha añadido la clase para guardar las métricas. 
* Se añadido objeto para manejar las peticiones con fetch.
* Se añade un callback opcional para medir las métricas.

### CODE REVIEW

File: packages/plugin/src/lib/api/api.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+export class ResponseError extends Error {
+  response: Response;
+
+  constructor(message: string, res: Response) {
+    super(message);
+    this.response = res;
+  }
+}
```
Comment (73nko) [inline packages/plugin/src/lib/api/api.ts]: We could create a specific file for this class, inside a utils folder. I think we are gonna reuse it and, in any case, it doesn't have too much sense to have an error class inside an API folder.

File: packages/plugin/src/lib/types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// eslint-disable-next-line @nrwl/nx/enforce-module-boundaries
+import { Metric } from 'web-vitals';
+
+export interface ProjectMetric extends Metric {
+  projectId: string;
+}
```
Comment (73nko) [inline packages/plugin/src/lib/types/index.d.ts]: What do you think if we create a global.d.ts file so we can have a global namespace to contain all these types and interfaces that are gonna be used along all the plugin?

File: packages/plugin/src/lib/plugin.ts [hunk 3]
```diff
@@ -1,5 +1,69 @@
...
+import { postMetric } from './api';
 
-export function plugin(): string {
-  return utils() + ' plugin';
+export interface PluginConfig {
+  id: string;
+  projectId: string;
+  urlEndpoint: string;
 }
+
```
Comment (73nko) [inline packages/plugin/src/lib/plugin.ts]: I think this interface should be allocated in our global namespace. We are probably gonna reuse it more times and it's a global part of the library.

File: packages/plugin/package-lock.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    "": {
+      "name": "@pulse-tracker/plugin",
+      "version": "0.0.1",
+      "dependencies": {
+        "web-vitals": "^3.3.1"
+      }
+    },
+    "node_modules/web-vitals": {
+      "version": "3.3.1",
+      "resolved": "https://inditex.jfrog.io/inditex/api/npm/node-public/web-vitals/-/web-vitals-3.3.1.tgz",
```
Comment (73nko) [inline packages/plugin/package-lock.json]: Be careful working from the work computer! 😆 
It was requiring a valid password to install the dependencies

File: packages/plugin/src/lib/api/api.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  return response.json().catch(() => ({}))
+}
+
+
+async function post<T, U>(path: string, body: T, config?: RequestInit): Promise<U> {
+  const init = {method: 'post', body: JSON.stringify(body), ...config}
+  return await http<U>(path, init)
+}
```
Comment (73nko) [inline packages/plugin/src/lib/api/api.ts:18]: Beautiful http handling! 🙌

File: packages/plugin/src/lib/plugin.ts [hunk 6]
```diff
@@ -1,5 +1,69 @@
...
+    onTTFB(callback);
+  }
+}
+
+export const plugin = {
+  config: (config: PluginConfig) => {
+    const plugin = new PluginRum(config);
+    return Object.freeze(plugin);
+  },
+};
```
Comment (73nko) [inline packages/plugin/src/lib/plugin.ts:61]: Love this! great idea the freeze pattern! 👍

File: packages/plugin/src/lib/plugin.ts [hunk 7]
```diff
@@ -1,5 +1,69 @@
...
-  return utils() + ' plugin';
+export interface PluginConfig {
+  id: string;
+  projectId: string;
+  urlEndpoint: string;
 }
+
+class PluginRum {
+  private _config: PluginConfig;
+  private _metrics: Record<string, Metric>;
```
Comment (73nko) [inline packages/plugin/src/lib/plugin.ts]: I don't think it's a good idea to save the metrics locally. This plugin is gonna be stored in a client application and we shall be as light as possible.

**Review by 73nko — COMMENTED** [review]
Great work! I've added a few not-important comments! But the class and implementation are perfect!

File: packages/plugin/package-lock.json (hunk 4 above)
Comment (jcantosg) [reply packages/plugin/package-lock.json]: Oops, what a mistake, I have already corrected it.

Thanks!.

File: packages/plugin/src/lib/api/api.ts (hunk 1 above)
Comment (jcantosg) [reply packages/plugin/src/lib/api/api.ts]: Done!!

File: packages/plugin/src/lib/types/index.d.ts (hunk 2 above)
Comment (jcantosg) [reply packages/plugin/src/lib/types/index.d.ts]: Done!!

File: packages/plugin/src/lib/plugin.ts (hunk 3 above)
Comment (jcantosg) [reply packages/plugin/src/lib/plugin.ts]: Done!!

File: packages/plugin/src/lib/plugin.ts (hunk 7 above)
Comment (jcantosg) [reply packages/plugin/src/lib/plugin.ts]: Done!!

**Review by 73nko — APPROVED** [review]

**Pull request merged**